invcache/invalidating_cache.go

```diff
@@ -0,0 +1,165 @@
+package invcache
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/goware/cachestore"
+	"github.com/goware/pubsub"
+)
+
+const (
+	ChannelID = "cache_invalidation"
+)
+
+type CacheInvalidationMessage struct {
+	Key    string `json:"key"`
+	Origin string `json:"origin"`
+}
+
+type InvalidatingCache[V any] struct {
+	store      cachestore.Store[V]
+	pubsub     pubsub.PubSub[CacheInvalidationMessage]
+	instanceID string
+}
+
+func NewInvalidatingCache[V any](store cachestore.Store[V], ps pubsub.PubSub[CacheInvalidationMessage], instanceID string) *InvalidatingCache[V] {
+	return &InvalidatingCache[V]{
+		store:      store,
+		pubsub:     ps,
+		instanceID: instanceID,
+	}
+}
+
+func (ic *InvalidatingCache[V]) Exists(ctx context.Context, key string) (bool, error) {
+	return ic.store.Exists(ctx, key)
+}
+
+func (ic *InvalidatingCache[V]) Get(ctx context.Context, key string) (V, bool, error) {
+	return ic.store.Get(ctx, key)
+}
+
+func (ic *InvalidatingCache[V]) GetEx(ctx context.Context, key string) (V, *time.Duration, bool, error) {
+	return ic.store.GetEx(ctx, key)
+}
+
+func (ic *InvalidatingCache[V]) BatchGet(ctx context.Context, keys []string) ([]V, []bool, error) {
+	return ic.store.BatchGet(ctx, keys)
+}
+
+func (ic *InvalidatingCache[V]) Set(ctx context.Context, key string, value V) error {
+	if err := ic.store.Set(ctx, key, value); err != nil {
+		return err
+	}
+	return ic.publishInvalidation(ctx, key)
+}
+
+func (ic *InvalidatingCache[V]) SetEx(ctx context.Context, key string, value V, ttl time.Duration) error {
+	if err := ic.store.SetEx(ctx, key, value, ttl); err != nil {
+		return err
+	}
+	return ic.publishInvalidation(ctx, key)
+}
+
+func (ic *InvalidatingCache[V]) BatchSet(ctx context.Context, keys []string, values []V) error {
+	if err := ic.store.BatchSet(ctx, keys, values); err != nil {
+		return err
+	}
+
+	var failedKeys []string
+	for _, k := range keys {
```

Comment:
we should pass it a slice of keys here instead of individual calls, as others mentioned in other parts of the code

Reply:
Done!